src/app/base.ts

```diff
@@ -138,62 +140,74 @@ export const CommonHeaders = [
-    "{task-id}": "9jBI6WHDwk60lEWpL4TQ92UAGLhh",
-    "{extension-id}": "com.contoso.roamingSettings",
-    "{host-name}": "cie493742.sharepoint.com",
-    "{server-relative-path}": "sites/Contoso/Operations/Manufacturing",
-    "{group-id-for-teams}": "d2b6c7fe-f440-446b-99d8-9ac12e036bf0",
-    "{channel-id}": "ee3f919c-28e1-4659-86d2-8e37e581335c",
-    "{today}": today.toISOString(),
-    "{next-week}": nextWeek.toISOString(),
-    "AUTHENTICATED_DOMAIN": () => {
-        try {
-            return AppComponent.explorerValues.authentication.user.emailAddress.split("@")[1];
-        } catch(e) {
-            return "example@contoso.com"
-        }
-    },
-    "FULL_USER_EMAIL": () => {
-        try {
-            return AppComponent.explorerValues.authentication.user.emailAddress;
-        } catch(e) {
-            return "example@contoso.com"
-        }
-    }
+];
+
+export interface Token {
+    placeholder: string
+
+    // base defaults to replace the placeholder with. Not used if any of the below are defined
+    defaultValue?: string
+    defaultValueFn?: Function
+
+    // when the user is not authenticated, use these values for the demo tenant
+    demoTenantValue?: string
+    demoTenantValueFn?: Function
+
+    // when the user is authenticated with MSA or AAD, replace token with these values
+    authenticatedUserValue?: string
+    authenticatedUserValueFn?: Function
 }
 
-export function substitueTokens(query:SampleQuery) {
-    for (let token in Tokens) {
-        if (query.requestUrl.indexOf(token) != -1) {
-            query.requestUrl = query.requestUrl.replace(token, Tokens[token]);
+/*
+ * Given a token, go through each of the possible replacement scenarios and find which value to
+ * replace the token with.
+ * Order: Authenticated user values, demo tenant replacament values, default replacement values.
+ */
+function getTokenSubstituteValue(token:Token) {
+    let priorityOrder = []; // desc
+
+    if (isAuthenticated()) {
+        priorityOrder.push(token.authenticatedUserValueFn);
+        priorityOrder.push(token.authenticatedUserValue);
+    } else {
+        priorityOrder.push(token.demoTenantValueFn);
+        priorityOrder.push(token.demoTenantValue);
+    }
+
+    priorityOrder.push(token.defaultValueFn);
+    priorityOrder.push(token.defaultValue);
+
+    for (let tokenVal of priorityOrder) {
+        if (!tokenVal) {
+            continue;
+        }
+        if (typeof tokenVal === "string") {
+            return tokenVal;
+        } else if (typeof tokenVal === "function") {
+            return tokenVal();
         }
     }
+
 }
 
-export function substituePostBodyTokens(query:SampleQuery) {
-    for (let token in Tokens) {
-        if (query.postBody.indexOf(token) != -1) {
-            let val;
-            if (typeof Tokens[token] == "string") {
-                val = Tokens[token];
-            } else {
-                val = Tokens[token]();
+export function substituteTokens(query:SampleQuery) {
```

Comment:
Comment what this function does for us. Unless you are committing to maintaining this forever :)